### ISSUE TITLE
no-restyle cannot classify named spacing scales (allow: ["spacing"] misses gap-sm, p-md)

### ISSUE THREAD

**Issue by refjamie** [shadcn-ui/lint#44]
### Summary

`no-restyle` classifies spacing utilities by parsing the class name against the numeric scale. On a project whose Tailwind v4 `@theme` defines a **named** spacing scale, the rule never recognises those utilities as spacing, so `allow: ["spacing"]` cannot cover them and there is no way to permit them short of listing every prefix as a glob.

### Reproduction

Theme (Tailwind v4, `app/globals.css`):

```css
@theme {
  --spacing-xs: 4px;
  --spacing-sm: 8px;
  --spacing-md: 16px;
  --spacing-lg: 24px;
  --spacing-xl: 32px;
  --spacing-2xl: 48px;
}
```

ESLint config:

```js
"shadcn/no-restyle": ["error", {
  allow: ["layout"],
  contracts: [{ pattern: "^CardContent$", allow: ["layout", "spacing"] }],
}]
```

Source:

```tsx
export function A() { return <CardContent className="gap-2">x</CardContent>; }
export function B() { return <CardContent className="gap-sm">x</CardContent>; }
export function C() { return <CardContent className="p-4">x</CardContent>; }
export function D() { return <CardContent className="p-md">x</CardContent>; }
```

### Actual

```
3:53  error  "gap-sm" is not allowed on <CardContent>: the grammar does not recognize it. Fix the spelling, or use a class Tailwind generates  shadcn/no-restyle
5:53  error  "p-md" is not allowed on <CardContent>: the grammar does not recognize it. Fix the spelling, or use a class Tailwind generates    shadcn/no-restyle
```

`gap-2` and `p-4` are correctly allowed by the contract. `gap-sm` and `p-md` are not.

### Expected

`gap-sm` and `p-md` are spacing utilities and should be allowed by `allow: ["spacing"]`, the same as `gap-2` and `p-4`.

They are also real classes that Tailwind generates. From the built CSS of the same project:

```css
--spacing-sm:8px
.gap-sm{gap:var(--spacing-sm)}
```

So the second half of the message, "use a class Tailwind generates", is misleading: Tailwind did generate it. The first half, "Fix the spelling", points at a typo that is not there, which is a confusing thing to hand a developer or an agent.

### Why it matters

Named spacing scales are common in projects inheriting a brand design system, where the token names come from the design side rather than from Tailwind's numeric ramp. In our case the whole scale is named, so **every** spacing utility in the codebase is unclassifiable: 31 findings across 22 files, none of which is a real restyle.

### Workaround

Listing the spacing prefixes as globs in the contract, which loses the semantics of `spacing` and has to be repeated per contract:

```js
contracts: [{
  pattern: "^Card(Content|Header|Title|Footer)$",
  allow: ["layout", "spacing", "gap-*", "space-x-*", "space-y-*",
          "p-*", "px-*", "py-*", "pt-*", "pb-*", "pl-*", "pr-*"],
}]
```

### Suggestion

Resolve spacing keys from the theme, the way component variants are already resolved from `components.json`, rather than from the numeric scale alone. Failing that, treating an unrecognised `<known-spacing-prefix>-<token>` as spacing rather than as a spelling mistake would fix both the classification and the message.

### Versions

`@shadcn/lint` 0.1.1 · eslint 9.39.5 · tailwindcss 4.3.3 · next 16.3.5 · eslint-config-next 16.3.5 · node 22

**Comment by aashish254** [comment]
Hi! I've put together a tested fix for this: when the grammar classifies a class as nothing, the project classifier now resolves `<spacing group>-<name>` against the theme's declared `--spacing-*` tokens (same theme-resolution pattern `namespaces.ts` already uses for color-shadowing namespaces).

Covers the exact repro from this issue (`gap-sm`, `p-md` with a named `@theme` scale), numeric classes are untouched, and undeclared names still report as unclassified. Includes 5 regression tests; full suite passes (746 tests).

My account currently can't open PRs to this org (looks like external PR creation is restricted), so the work is on my fork for now: https://github.com/aashish254/lint/tree/fix/named-spacing-scale-classification — happy to open the PR as soon as that's possible, or a maintainer can pull the branch directly if you'd like it sooner.

**Comment by Sebastian-Sole** [comment]
I tested #52 against a Tailwind v4 theme generated from Figma, and it does fix the case in this issue. On `50221d7`, `gap-2xl`, `p-2xl`, `m-s`, `h-control-md`, `min-h-control-2xl`, `size-icon-md`, `max-w-max-content-width` and `rounded-8` all classify correctly. On `main` (`a89d047`) every one of them is unclassified.

Two things in the same family survive it, so I want to flag them before this issue gets closed.

`border-sm` (from `--border-width-sm`) and `grid-cols-layout` (from `--grid-template-columns-layout`) are real classes that Tailwind 4.3.3 generates, but `cn` 0.4.0's `defaultConfig().theme` has 19 scales and neither `border-width` nor `grid-template-columns` is among them. So the `NAMESPACES` list in `theme-scales.ts` has nothing to merge those into. That fix starts in `cn`, not here.

`border-sm` is the worse of the two. It is not merely unclassified, it is reported as a color. That is the inversion that makes a category allowlist backfire: a project writing `allow: ["layout"]` still gets `grid-cols-layout` rejected, and `border-sm` is grouped with the colors rather than with borders.

How I checked: compiled each candidate against tailwindcss 4.3.3 and diffed the `@layer utilities` output to confirm the class is real, then read the group through `projectClassifierFor(page).groupOf(token)` on both branches.

**Comment by gerentegrok-cpu** [comment]
quem sao voces\????
